File: BlackFireIM/Example/MSIMSDK-UIKit/ChatDetail/Model/MSVoiceMessageCellData.m
//
//  MSVoiceMessageCellData.m
//  BlackFireIM
//
//  Created by benny wang on 2021/4/22.
//

#import "MSVoiceMessageCellData.h"
#import <AVFoundation/AVFoundation.h>
#import "MSHeader.h"
#import "NSFileManager+filePath.h"
#import "MSIMSDK.h"


@interface MSVoiceMessageCellData()<AVAudioPlayerDelegate>

@property(nonatomic,strong) AVAudioPlayer *audioPlayer;

@property(nonatomic,copy) NSString *wavPath;

@end
@implementation MSVoiceMessageCellData

- (instancetype)initWithDirection:(TMsgDirection)direction
{
    self = [super initWithDirection:direction];
    if (self) {
        if (direction == MsgDirectionIncoming) {
            _voiceImage = [UIImage bf_imageNamed:@"receiver_voice"];
            _voiceAnimationImages = @[[UIImage bf_imageNamed:@"receiver_voice_play_1"],
                                      [UIImage bf_imageNamed:@"receiver_voice_play_2"],
                                      [UIImage bf_imageNamed:@"receiver_voice_play_3"]];
        } else {
            _voiceImage = [UIImage bf_imageNamed:@"sender_voice"];
            _voiceAnimationImages = @[[UIImage bf_imageNamed:@"sender_voice_play_1"],
                                      [UIImage bf_imageNamed:@"sender_voice_play_2"],
                                      [UIImage bf_imageNamed:@"sender_voice_play_3"]];
        }
    }

    return self;
}

- (MSIMVoiceElem *)voiceElem
{
    return (MSIMVoiceElem *)self.elem;
}

- (CGSize)contentSize
{
    CGFloat bubbleWidth = TVoiceMessageCell_Back_Width_Min + self.voiceElem.duration / TVoiceMessageCell_Max_Duration * Screen_Width;
    if(bubbleWidth > TVoiceMessageCell_Back_Width_Max){
        bubbleWidth = TVoiceMessageCell_Back_Width_Max;
    }

    CGFloat bubbleHeight = TVoiceMessageCell_Duration_Size.height;
    if (self.direction == MsgDirectionIncoming) {
        bubbleWidth = MAX(bubbleWidth, [MSBubbleMessageCellData incommingBubble].size.width);
        bubbleHeight = 40;
    } else {
        bubbleWidth = MAX(bubbleWidth, [MSBubbleMessageCellData outgoingBubble].size.width);
        bubbleHeight = 40;
    }
    return CGSizeMake(bubbleWidth+TVoiceMessageCell_Duration_Size.width, bubbleHeight);
}

- (void)playVoiceMessage
{
    if (self.isPlaying) {
        return;
    }
    self.isPlaying = YES;
    NSString *path = self.voiceElem.path;
    if (path && [[NSFileManager defaultManager]fileExistsAtPath:path]) {
        [self playInternal:path];
    }else {
        if (self.isDownloading) {
            return;
        }
        //下载
        self.isDownloading = YES;
        WS(weakSelf)
        NSString *savePath = [[NSFileManager pathForIMVoice] stringByAppendingPathComponent:[self.voiceElem.url lastPathComponent]];
        
        [[MSIMManager sharedInstance].uploadMediator ms_downloadFromUrl:self.voiceElem.url toSavePath:savePath progress:^(CGFloat progress) {
                    
                } succ:^(NSString * _Nonnull url) {
                    
                    weakSelf.isDownloading = NO;
                    weakSelf.voiceElem.path = savePath;
                    [weakSelf playInternal:savePath];
                    
                } fail:^(NSInteger code, NSString * _Nonnull desc) {
                    weakSelf.isDownloading = NO;
                    [weakSelf stopVoiceMessage];
        }];
    }
}

- (void)playInternal:(NSString *)path
{
    if (!self.isPlaying) return;

    [[AVAudioSession sharedInstance] setCategory:AVAudioSessionCategoryPlayback error:nil];
    NSURL *url = [NSURL fileURLWithPath:path];
    [self.audioPlayer stop];
    self.audioPlayer = [[AVAudioPlayer alloc]initWithData:[NSData dataWithContentsOfURL:url] error:nil];
    self.audioPlayer.delegate = self;
    BOOL result = [self.audioPlayer play];
    if (!result) {
        self.isPlaying = NO;
        [MSHelper showToastFail:@"音频文件不存在或已损坏"];
    }
}

- (void)stopVoiceMessage
{
    if ([self.audioPlayer isPlaying]) {
        [self.audioPlayer stop];
        self.audioPlayer = nil;
    }
    self.isPlaying = NO;
}

- (void)audioPlayerDidFinishPlaying:(AVAudioPlayer *)player successfully:(BOOL)flag;
{
    self.isPlaying = NO;
    [[NSFileManager defaultManager] removeItemAtPath:self.wavPath error:nil];
}

- (void)audioPlayerDecodeErrorDidOccur:(AVAudioPlayer *)player error:(NSError * __nullable)error
{
    self.isPlaying = NO;
    MSLog(@"音频播放失败：%@",error);
    [MSHelper showToastFail:@"音频文件解码失败"];
}

- (NSString *)reuseId
{
    return TVoiceMessageCell_ReuseId;
}

@end

File: Swift/MSIM-Swift/Lib/MSIMSDK-UIKit/ChatDetail/Model/MSVoiceMessageCellData.swift
//
//  MSVoiceMessageCellData.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/8/5.
//

import UIKit
import MSIMSDK
import AVFoundation


open class MSVoiceMessageCellData: MSBubbleMessageCellData {

    public var voiceAnimationImages: [UIImage] = []
    
    public var voiceImage: UIImage!
    
    @objc dynamic public var isPlaying: Bool = false
    
    public private(set) var isDownloading: Bool = false
    
    public func playVoiceMessage() {
        if isPlaying {return}
        isPlaying = true
        guard let path = message.voiceElem!.path else {return}
        if FileManager.default.fileExists(atPath: path) {
            playInternal(path: path)
        }else {
            if isDownloading {return}
            if message.voiceElem?.url == nil {return}
            isDownloading = true
            let savePath = FileManager.pathForIMVoice().appendingFormat("/%@", (message.voiceElem!.url! as NSString).lastPathComponent)
            MSIMManager.sharedInstance().uploadMediator?.ms_download?(fromUrl: message.voiceElem!.url!, toSavePath: savePath, progress: { progress in
                
            }, succ: { url in
                
                self.isDownloading = false
                self.message.voiceElem?.path = savePath
                self.playInternal(path: savePath)
                
            }, fail: { code, desc in
                self.isDownloading = false
                self.stopVoiceMessage()
            })
        }
    }
    
    public func stopVoiceMessage() {
        if audioPlayer?.isPlaying == true {
            audioPlayer?.stop()
            audioPlayer = nil
        }
        isPlaying = false
    }
    
    private var audioPlayer: AVAudioPlayer?
    
    public override init(direction: TMsgDirection) {
        super.init(direction: direction)
        if direction == .inComing {
            voiceImage = UIImage.bf_imageNamed(name: "receiver_voice")
            voiceAnimationImages = [UIImage.bf_imageNamed(name: "receiver_voice_play_1")!,UIImage.bf_imageNamed(name: "receiver_voice_play_2")!,UIImage.bf_imageNamed(name: "receiver_voice_play_3")!]
        }else {
            voiceImage = UIImage.bf_imageNamed(name: "sender_voice")
            voiceAnimationImages = [UIImage.bf_imageNamed(name: "sender_voice_play_1")!,UIImage.bf_imageNamed(name: "sender_voice_play_2")!,UIImage.bf_imageNamed(name: "sender_voice_play_3")!]
        }
    }
    
    public override func contentSize() -> CGSize {
        var bubbleWidth = MSMcros.TVoiceMessageCell_Back_Width_Min + CGFloat(message.voiceElem!.duration) / MSMcros.TVoiceMessageCell_Max_Duration * UIScreen.width
        if bubbleWidth > MSMcros.TVoiceMessageCell_Back_Width_Max {
            bubbleWidth = MSMcros.TVoiceMessageCell_Back_Width_Max
        }
        
        var bubbleHeight = MSMcros.TVoiceMessageCell_Duration_Size.height
        if direction == .inComing {
            bubbleWidth = max(bubbleWidth, incommingBubble.size.width)
            bubbleHeight = 40
        }else {
            bubbleWidth = max(bubbleWidth, outgoingBubble.size.width)
            bubbleHeight = 40
        }
        return CGSize(width: bubbleWidth + MSMcros.TVoiceMessageCell_Duration_Size.width, height: bubbleHeight)
    }
    
    public override var reUseId: String {
        return MSMcros.TVoiceMessageCell_ReuseId
    }
}

private extension MSVoiceMessageCellData {
    
    func playInternal(path: String) {
        
        if !isPlaying {return}
        
        try? AVAudioSession.sharedInstance().setCategory(.playback, options: [])
        let url = URL(fileURLWithPath: path)
        audioPlayer?.stop()
        do {
            audioPlayer = try AVAudioPlayer(data: Data(contentsOf: url))
            audioPlayer?.delegate = self
            if audioPlayer?.play() == false {
                isPlaying = false
                MSHelper.showToastFailWithText(text: "音频文件不存在或已损坏")
            }
        }catch {
            print(error)
            isPlaying = false
        }
    }
}

extension MSVoiceMessageCellData: AVAudioPlayerDelegate {
    
    public func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
        isPlaying = false
    }
    
    public func audioPlayerDecodeErrorDidOccur(_ player: AVAudioPlayer, error: Error?) {
        isPlaying = false
        print("音频播放失败：%@",error!)
        MSHelper.showToastFailWithText(text: "音频文件解码失败")
    }
}
